### DISCUSSION TITLE
How to destroy the controller

### DISCUSSION THREAD

**Discussion by david-dot-krupicka** [mojolicious/mojo#2137] [Q&A]
Hello,

please help.
In our tests, we are instantiating the mocked controller like this:
```
    $self->{ app } = Test::MockObject::Extends->new(MyApp->new);
    $self->{ tx } = Mojo::Transaction::HTTP->new;

    my $c = $class->new(app => $self->{app}, tx => $self->{tx});
```
... where `$class` inherits from the base controller.

Internally, this opens a filehandle to the log, creates a database connection, etc.
How to properly destroy it? I could write a teardown method to close the handle, close the connection, and undef the controller, but I'm not sure if it'd be complete.

**Comment by christopherraa** [comment]
The somewhat flippant answer to "_how to properly destroy it_" depends on what you do inside it. If you do anything odd (creating dangling references / messing with refcounts) then you might have to take care to implement proper cleanup yourself. Doing leak-testing will tell you if this is required. Though, for most use cases nothing fancy is requires. `undef`-ing objects should do what you want.

Looking at the provided code I do wonder if you might be overcomplicating something. I do not think I have ever needed to mock a controller, and we do _heavy_ testing at $work. Usually we just instantiate our apps for testing using `Test::Mojo` and enjoy that Mojo just does everything exactly like we expect. The few times when mocking is needed it is usually for external modules where we have little or no influence and need to manipulate something to simulate various external failure states.
